File: backend/convert_excel.js
const XLSX = require('xlsx');
const fs = require('fs');
const path = require('path');

const SRC = 'c:/temp/Table SSPOSBIZ.xlsx';
const OUT = path.join(__dirname, '..', 'scripts', 'data.json');

function excelDateToISO(v) {
  if (v == null || v === '') return null;
  if (v instanceof Date && !isNaN(v)) {
    return v.toISOString().slice(0, 10);
  }
  if (typeof v === 'number') {
    const d = new Date(Math.round((v - 25569) * 86400 * 1000));
    return isNaN(d) ? null : d.toISOString().slice(0, 10);
  }
  const s = String(v).trim();
  if (!s) return null;
  return s;
}

function toStr(v) {
  if (v == null) return null;
  const s = String(v).trim();
  return s || null;
}

function toInt(v) {
  if (v == null) return null;
  const s = String(v).trim();
  if (!s) return null;
  const n = Number(s);
  return isNaN(n) ? null : Math.round(n);
}

function toBool(v) {
  if (v == null) return null;
  const s = String(v).trim();
  if (!s) return null;
  return /^1|true|oui$/i.test(s) ? 1 : 0;
}

const wb = XLSX.readFile(SRC, { cellDates: true });
const ws = wb.Sheets['A'];
const rows = XLSX.utils.sheet_to_json(ws, { header: 1, raw: true, defval: null });

const customers = [];
for (let i = 1; i < rows.length; i++) {
  const r = rows[i] || [];
  if (!r[0] && !r[1]) continue; // ligne vide
  customers.push({
    customer: toStr(r[0]),
    customer_site: toStr(r[1]),
    date_end_licence: excelDateToISO(r[2]),
    product: toStr(r[3]),
    mac_address: toStr(r[4]),
    lic_35: toInt(r[5]),
    siel_id: toStr(r[6]),
    registered_company: toStr(r[7]),
    number_user: toInt(r[8]),
    contract: toBool(r[9]),
    last_lac: toStr(r[10]),
    date_last_lac: excelDateToISO(r[11]),
    udl_contract: toStr(r[12]),
    dlu_contract: toStr(r[13]),
    info_divers: toStr(r[14])
  });
}

const data = {
  customers,
  meta: {
    source_file: SRC,
    customer_count: customers.length
  }
};

fs.writeFileSync(OUT, JSON.stringify(data, null, 2), 'utf-8');
console.log(`OK: ${customers.length} clients importés depuis ${SRC}`);
console.log(`Fichier écrit: ${OUT}`);

// Afficher 3 exemples pour vérification
for (const c of customers.slice(0, 3)) {
  console.log(JSON.stringify(c));
}

File: scripts/convert_excel.py
import pandas as pd
import json
import sys

src = r"c:\temp\SSP Openscape Business.xlsx"
out = "C:\\Users\\didie\\OneDrive\\Documents\\Default Project\\ssp-app\\scripts\\data.json"

def clean_date(v):
    if v is None or (isinstance(v, float) and v != v):  # NaN
        return None
    if pd.isna(v):
        return None
    if hasattr(v, 'strftime'):
        return v.strftime('%Y-%m-%d')
    return str(v)

def clean_str(v):
    if v is None:
        return None
    if isinstance(v, float) and v != v:
        return None
    if pd.isna(v):
        return None
    s = str(v).strip()
    return s if s else None

def to_int(v):
    if v is None:
        return None
    if isinstance(v, float) and v != v:
        return None
    if pd.isna(v):
        return None
    if isinstance(v, str):
        v = v.replace(',', '').strip()
        if v.upper() == 'PAYGO':
            return None
        try:
            return int(float(v))
        except:
            return None
    return int(v)

xl = pd.ExcelFile(src)

# --- Feuille principale ---
df = pd.read_excel(xl, sheet_name="SSP ALL CUSTOMER", header=0)
customers = []
for _, r in df.iterrows():
    customers.append({
        "customer": clean_str(r.get("Customer")),
        "customer_site": clean_str(r.get("Customer site")),
        "date_end_licence": clean_date(r.get("Date_end_licence")),
        "product": clean_str(r.get("Product")),
        "mac_address": clean_str(r.get("Mac_address")),
        "siel_id": clean_str(r.get("SIEL ID")),
        "registered_company": clean_str(r.get("Registered Compagny")),
        "number_user": to_int(r.get("Number User")),
        "contract": to_int(r.get("Contract")),
        "remote": to_int(r.get("Remote")),
        "last_lac": clean_str(r.get("LAST LAC")),
        "date_last_lac": clean_date(r.get("DATE LAST LAC")),
        "contrat_damovo": clean_str(r.get("Contrat Damovo")),
        "info_divers": clean_str(r.get("INFORMATION DIVERS"))
    })

# --- Feuille LAST SSP ---
df2 = pd.read_excel(xl, sheet_name="LAST SSP", header=0)
last_ssp = []
for _, r in df2.iterrows():
    last_ssp.append({
        "lac": clean_str(r.get("LAC")),
        "total_quantity": to_int(r.get("Total Quantity")),
        "available_quantity": to_int(r.get("Available Quantity")),
        "feature": clean_str(r.get("Feature")),
        "create_date": clean_date(r.get("Create Date")),
        "siel_id": clean_str(r.get("SIEL-ID"))
    })

# --- Feuille Par Mois ---
df3 = pd.read_excel(xl, sheet_name="Par Mois", header=None)
par_mois = df3.values.tolist()

data = {
    "customers": customers,
    "last_ssp": last_ssp,
    "par_mois": par_mois,
    "meta": {
        "source_file": src,
        "customer_count": len(customers)
    }
}

with open(out, "w", encoding="utf-8") as f:
    json.dump(data, f, ensure_ascii=False, indent=2)

print(f"OK: {len(customers)} customers, {len(last_ssp)} last_ssp, {len(par_mois)} par_mois lignes")
print(f"Fichier écrit: {out}")
